timemachine/fe/single_topology_v3.py

```diff
@@ -0,0 +1,539 @@
+import warnings
+from collections.abc import Iterable
+
+import numpy as np
+
+from timemachine.fe import system, topology, utils
+from timemachine.fe.dummy import identify_dummy_groups, identify_root_anchors
+from timemachine.lib import potentials
+
+
+class MultipleAnchorWarning(UserWarning):
+    pass
+
+
+class CoreBondChangeWarning(UserWarning):
+    pass
+
+
+class MissingAngleError(RuntimeError):
+    pass
+
+
+def get_romol_bonds(mol):
+    """
+    Return bond idxs given a mol. These are not canonicalized.
+    """
+    bond_list = []
+    for bond in mol.GetBonds():
+        bond_list.append([bond.GetBeginAtomIdx(), bond.GetEndAtomIdx()])
+    return bond_list
+
+
+def get_neighbors(bond_idxs, query):
+    """
+    Get neighbors of query atom query.
+    """
+    nbs = set()
+    for i, j in bond_idxs:
+        if i == query:
+            nbs.add(j)
+        if j == query:
+            nbs.add(i)
+    return nbs
+
+
+def recursive_map(items, mapping):
+    """recursively replace items in a list of tuple
+    mapping = np.arange(100)[::-1]
+    items = [[0,2,3], [5,1,[2,5,6]], 3]
+    result = recursive_map(items, mapping)
+    """
+    if isinstance(items, Iterable):
+        res = []
+        for item in items:
+            res.append(recursive_map(item, mapping))
+        return tuple(res)
+    else:
+        return mapping[items]
+
+
+def find_nbs(mol, atom):
+    nbs = [a.GetIdx() for a in mol.GetAtomWithIdx(atom).GetNeighbors()]
+    return nbs
+
+
+def setup_dummy_interactions_from_ff(ff, mol, dummy_group, root_anchor_atom, nbor_anchor_atom):
+    """
+    Setup interactions involving atoms in a given dummy group.
+    """
+    top = topology.BaseTopology(mol, ff)
+
+    bond_params, hb = top.parameterize_harmonic_bond(ff.hb_handle.params)
+    angle_params, ha = top.parameterize_harmonic_angle(ff.ha_handle.params)
+    improper_params, it = top.parameterize_improper_torsion(ff.it_handle.params)
+
+    bond_idxs = hb.get_idxs()
+    angle_idxs = ha.get_idxs()
+    improper_idxs = it.get_idxs()
+
+    return setup_dummy_interactions(
+        bond_idxs,
+        bond_params,
+        angle_idxs,
+        angle_params,
+        improper_params,
+        improper_idxs,
+        dummy_group,
+        root_anchor_atom,
+        nbor_anchor_atom,
+    )
+
+
+def setup_dummy_interactions(
+    bond_idxs,
+    bond_params,
+    angle_idxs,
+    angle_params,
+    improper_idxs,
+    improper_params,
+    dummy_group,
+    root_anchor_atom,
+    nbor_anchor_atom,
+):
```

Comment:
Is there a reason this is a separate function?

Reply:
This may be useful later on if we want to `setup_dummy_interactions` with already parameterized terms - the `_ff` version is just a utility. It's also more clear to document the required terms here explicitly (eg. bond, angle, impropers only).

Reply:
sgtm